File: packages/domain/src/rbac.ts
export const ADMIN_ROLES = [
  'SUPER_ADMIN',
  'HQ_ADMIN',
  'BRANCH_MANAGER',
  'FRONT_DESK',
  'COACH',
  'FINANCE',
] as const;
export type AdminRole = (typeof ADMIN_ROLES)[number];

/** Users are assigned roles, but everything internal checks fine-grained permissions. */
export const PERMISSIONS = [
  'dashboard.view',
  'members.view',
  'members.manage',
  'members.adjust_credits',
  'ledger.view',
  'ledger.reverse',
  'operations.view',
  'class_types.manage',
  'sessions.manage',
  'coaches.manage',
  'bookings.manage',
  'attendance.manage',
  'access.view',
  'access.simulate',
  'commercial.view',
  'packages.manage',
  'payments.view',
  'payments.simulate',
  'vouchers.manage',
  'refunds.manage',
  'engagement.view',
  'campaigns.manage',
  'reports.view',
  'reports.financial',
  'incentives.view',
  'incentives.manage',
  'config.view',
  'branches.manage',
  'gates.manage',
  'users.manage',
  'rules.update',
  // The people side. Employee records carry home addresses, bank accounts and
  // next of kin, so 'hris.view' is a narrower grant than it looks.
  'hris.view',
  'hris.manage',
  'hris.attendance',
  'hris.approve',
  // Stock. Looking is broad — the front desk has to know whether a shirt is on
  // the shelf — but changing a quantity is not.
  'inventory.view',
  'inventory.manage',
  'inventory.count',
  // Buying. Raising, approving and receiving are three separate grants,
  // because one person doing all three is how invoices get paid for goods that
  // never arrived.
  'purchasing.view',
  'purchasing.manage',
  'purchasing.approve',
  'purchasing.receive',
  // Paying a supplier is a finance act, not a buying one: the person who
  // chooses what to order should not also move the money out.
  'purchasing.pay',
  // The till. Selling is the counter's job; unwinding a paid sale is not.
  'pos.view',
  'pos.sell',
  'pos.manage',
  'pos.void',
  // Loyalty. Handing out points by hand is its own grant, for the same reason
  // adjusting somebody's credits is.
  'crm.view',
  'crm.manage',
  'crm.approve',
  'crm.adjust',
] as const;
export type Permission = (typeof PERMISSIONS)[number];

const ALL: readonly Permission[] = PERMISSIONS;

export const ROLE_PERMISSIONS: Record<AdminRole, readonly Permission[]> = {
  SUPER_ADMIN: ALL,
  HQ_ADMIN: ALL.filter((p) => p !== 'users.manage' && p !== 'rules.update'),
  BRANCH_MANAGER: [
    'dashboard.view',
    'members.view',
    'members.manage',
    'ledger.view',
    'operations.view',
    'class_types.manage',
    'sessions.manage',
    'coaches.manage',
    'bookings.manage',
    'attendance.manage',
    'access.view',
    'access.simulate',
    'reports.view',
    'incentives.view',
    'config.view',
    'hris.view',
    'hris.attendance',
    'hris.approve',
    'inventory.view',
    'inventory.manage',
    'inventory.count',
    // A branch manager is the first signature on the chain, and the person who
    // signs for goods arriving at their own branch.
    'purchasing.view',
    'purchasing.manage',
    'purchasing.approve',
    'purchasing.receive',
    'pos.view',
    'pos.sell',
    'pos.manage',
    'pos.void',
    'crm.view',
  ],
  FRONT_DESK: [
    'dashboard.view',
    'members.view',
    'ledger.view',
    'operations.view',
    'bookings.manage',
    'attendance.manage',
    'access.view',
    'access.simulate',
    'commercial.view',
    'payments.view',
    // They run the till and answer "have you got this in a medium", so they
    // sell and they look. Voiding a paid sale needs a manager, and changing a
    // stock figure is not a counter job.
    'pos.view',
    'pos.sell',
    'inventory.view',
    'crm.view',
  ],
  COACH: ['dashboard.view', 'operations.view', 'attendance.manage', 'members.view'],
  FINANCE: [
    'dashboard.view',
    'members.view',
    'ledger.view',
    'commercial.view',
    'payments.view',
    'payments.simulate',
    'refunds.manage',
    'reports.view',
    'reports.financial',
    'incentives.view',
    'incentives.manage',
    'hris.view',
    // Finance is the second signature on a purchase and the reason stock
    // valuation exists, but never touches a quantity or a till.
    'inventory.view',
    'purchasing.view',
    'purchasing.approve',
    // And the only role besides the super admin that may actually pay one.
    'purchasing.pay',
    'pos.view',
    'crm.view',
  ],
};

export function hasPermission(role: AdminRole, permission: Permission): boolean {
  return ROLE_PERMISSIONS[role].includes(permission);
}

File: apps/backend/internal/domain/rbac.go
package domain

// AdminRole is a staff account's role. Roles are coarse; everything internal
// checks a fine-grained permission so the matrix can change without touching
// call sites.
type AdminRole string

const (
	RoleSuperAdmin    AdminRole = "SUPER_ADMIN"
	RoleHQAdmin       AdminRole = "HQ_ADMIN"
	RoleBranchManager AdminRole = "BRANCH_MANAGER"
	RoleFrontDesk     AdminRole = "FRONT_DESK"
	RoleCoach         AdminRole = "COACH"
	RoleFinance       AdminRole = "FINANCE"
)

var AdminRoles = []AdminRole{
	RoleSuperAdmin, RoleHQAdmin, RoleBranchManager, RoleFrontDesk, RoleCoach, RoleFinance,
}

// Permission names a single capability. The string values are part of the API
// contract: the admin panel receives the caller's list at login and uses it to
// decide what to render.
type Permission string

const (
	PermDashboardView       Permission = "dashboard.view"
	PermMembersView         Permission = "members.view"
	PermMembersManage       Permission = "members.manage"
	PermMembersAdjustCredit Permission = "members.adjust_credits"
	PermLedgerView          Permission = "ledger.view"
	PermLedgerReverse       Permission = "ledger.reverse"
	PermOperationsView      Permission = "operations.view"
	PermClassTypesManage    Permission = "class_types.manage"
	PermSessionsManage      Permission = "sessions.manage"
	PermCoachesManage       Permission = "coaches.manage"
	PermBookingsManage      Permission = "bookings.manage"
	PermAttendanceManage    Permission = "attendance.manage"
	PermAccessView          Permission = "access.view"
	PermAccessSimulate      Permission = "access.simulate"
	PermCommercialView      Permission = "commercial.view"
	PermPackagesManage      Permission = "packages.manage"
	PermPaymentsView        Permission = "payments.view"
	PermPaymentsSimulate    Permission = "payments.simulate"
	PermVouchersManage      Permission = "vouchers.manage"
	PermRefundsManage       Permission = "refunds.manage"
	PermEngagementView      Permission = "engagement.view"
	PermCampaignsManage     Permission = "campaigns.manage"
	PermReportsView         Permission = "reports.view"
	PermReportsFinancial    Permission = "reports.financial"
	PermIncentivesView      Permission = "incentives.view"
	PermIncentivesManage    Permission = "incentives.manage"
	PermConfigView          Permission = "config.view"
	PermBranchesManage      Permission = "branches.manage"
	PermGatesManage         Permission = "gates.manage"
	PermUsersManage         Permission = "users.manage"
	PermRulesUpdate         Permission = "rules.update"
	// HRIS is staff administration: the people who run the studio rather than
	// the members who train in it.
	PermHRISView       Permission = "hris.view"
	PermHRISManage     Permission = "hris.manage"
	PermHRISAttendance Permission = "hris.attendance"
	PermHRISApprove    Permission = "hris.approve"
)

// Permissions is the full list, in the order the admin panel expects.
var Permissions = []Permission{
	PermDashboardView, PermMembersView, PermMembersManage, PermMembersAdjustCredit,
	PermLedgerView, PermLedgerReverse, PermOperationsView, PermClassTypesManage,
	PermSessionsManage, PermCoachesManage, PermBookingsManage, PermAttendanceManage,
	PermAccessView, PermAccessSimulate, PermCommercialView, PermPackagesManage,
	PermPaymentsView, PermPaymentsSimulate, PermVouchersManage, PermRefundsManage,
	PermEngagementView, PermCampaignsManage, PermReportsView, PermReportsFinancial,
	PermIncentivesView, PermIncentivesManage, PermConfigView, PermBranchesManage,
	PermGatesManage, PermUsersManage, PermRulesUpdate,
	PermHRISView, PermHRISManage, PermHRISAttendance, PermHRISApprove,
}

// RolePermissions is the authorization matrix. Super Admin holds everything;
// HQ Admin holds everything except the two that reshape the system itself
// (staff accounts and business rules).
var RolePermissions = map[AdminRole][]Permission{
	RoleSuperAdmin: Permissions,
	RoleHQAdmin:    without(Permissions, PermUsersManage, PermRulesUpdate),
	RoleBranchManager: {
		PermDashboardView, PermMembersView, PermMembersManage, PermLedgerView,
		PermOperationsView, PermClassTypesManage, PermSessionsManage, PermCoachesManage,
		PermBookingsManage, PermAttendanceManage, PermAccessView, PermAccessSimulate,
		PermReportsView, PermIncentivesView, PermConfigView,
		PermHRISView, PermHRISAttendance, PermHRISApprove,
	},
	RoleFrontDesk: {
		PermDashboardView, PermMembersView, PermLedgerView, PermOperationsView,
		PermBookingsManage, PermAttendanceManage, PermAccessView, PermAccessSimulate,
		PermCommercialView, PermPaymentsView,
	},
	RoleCoach: {
		PermDashboardView, PermOperationsView, PermAttendanceManage, PermMembersView,
	},
	RoleFinance: {
		PermDashboardView, PermMembersView, PermLedgerView, PermCommercialView,
		PermPaymentsView, PermPaymentsSimulate, PermRefundsManage, PermReportsView,
		PermReportsFinancial, PermIncentivesView, PermIncentivesManage,
		PermHRISView,
	},
}

// HasPermission is the single authorization check in the system.
func HasPermission(role AdminRole, permission Permission) bool {
	return contains(RolePermissions[role], permission)
}

// PermissionsFor returns the caller's capability list, sent at login.
func PermissionsFor(role AdminRole) []Permission {
	perms, ok := RolePermissions[role]
	if !ok {
		return []Permission{}
	}
	out := make([]Permission, len(perms))
	copy(out, perms)
	return out
}

// IsAdminRole validates a role string arriving from a request.
func IsAdminRole(value string) bool {
	return contains(AdminRoles, AdminRole(value))
}

func without(all []Permission, excluded ...Permission) []Permission {
	out := make([]Permission, 0, len(all))
	for _, p := range all {
		if !contains(excluded, p) {
			out = append(out, p)
		}
	}
	return out
}
